File: app/src/main/java/albums/challenge/Application.java
package albums.challenge;

import albums.challenge.models.Results;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.ModelAndView;

import java.util.List;

@SpringBootApplication
@RestController
@EnableCaching
public class Application {
    @Autowired
    DataService dataService;

    @Autowired
    SearchService searchService;

    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }

    @GetMapping("/")
    public ModelAndView index() {
        var modelAndView = new ModelAndView();
        modelAndView.setViewName("index.html");
        return modelAndView;
    }

    @GetMapping("/api/search")
    public Results search(
            @RequestParam String query,
            @RequestParam(defaultValue = "") List<String> year,
            @RequestParam(defaultValue = "") List<String> price
    ) {
        return searchService.search(
                dataService.fetch(),
                query,
                year,
                price
        );
    }
}

File: src/main/scala/albums/challenge/Application.scala
package albums.challenge

import albums.challenge.models.Results
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.cache.annotation.EnableCaching
import org.springframework.web.bind.annotation.{GetMapping, RequestParam, RestController}
import org.springframework.web.servlet.ModelAndView

import scala.jdk.CollectionConverters.CollectionHasAsScala

@SpringBootApplication
@RestController
@EnableCaching
class Application @Autowired() (
    dataService: DataService,
    searchService: SearchService,
) {

  @GetMapping(Array("/"))
  def index(): ModelAndView = {
    val modelAndView = new ModelAndView()
    modelAndView.setViewName("index.html")
    modelAndView
  }

  @GetMapping(Array("/api/search"))
  def search(
      @RequestParam query: String,
      @RequestParam(defaultValue = "") year: java.util.List[String],
      @RequestParam(defaultValue = "") price: java.util.List[String],
  ): Results = {
    searchService.search(
      dataService.fetch(),
      query,
      year.asScala.toList,
      price.asScala.toList,
    )
  }
}

object Application extends App {
  SpringApplication.run(classOf[Application], args: _*)
}
